### PR TITLE
Ray tracing support for VR

### PR DESCRIPTION
**Pull request by fabien-unity**
### Purpose of this PR
Support single-pass instancing with ray tracing code

---
### Testing status
**Yamato Tests**: [passed](https://yamato.cds.internal.unity3d.com/jobs/78-ScriptableRenderPipeline/tree/hdrp-xr-rtx-pass1/.yamato%252Fupm-ci-hdrp.yml%2523All_HDRP_Tests/398210/job)

**Manual Tests**: What did you do?
- [x] Opened test project + Run graphic tests locally
- [x] Built a player

---
### Overall Product Risks
**Technical Risk**: Low

**Halo Effect**: Low

---
### Comments to reviewers
Recap of the changes:
- update Z dimension of dispatch calls to handle all views for VR
- update shaders to handle texture array correctly with VR

### CODE REVIEW

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipeline.cs [hunk 1]
```diff
@@ -1966,7 +1966,9 @@ void Callback(CommandBuffer c, HDCamera cam)
 
                 using (new ProfilingSample(cmd, "Render screen space shadows", CustomSamplerId.ScreenSpaceShadows.GetSampler()))
                 {
+                    hdCamera.xr.StartSinglePass(cmd, camera, renderContext);
```
Comment (anisunity) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipeline.cs]: What does this do? @fabien-unity

File: com.unity.render-pipelines.high-definition/Runtime/ShaderLibrary/TextureXR.hlsl [hunk 2]
```diff
@@ -61,18 +64,19 @@
...
     #define LOAD_TEXTURE2D_X_MSAA(textureName, unCoord2, sampleIndex)        LOAD_TEXTURE2D_ARRAY_MSAA(textureName, unCoord2, SLICE_ARRAY_INDEX, sampleIndex)
     #define LOAD_TEXTURE2D_X_LOD(textureName, unCoord2, lod)                 LOAD_TEXTURE2D_ARRAY_LOD(textureName, unCoord2, SLICE_ARRAY_INDEX, lod)
     #define SAMPLE_TEXTURE2D_X(textureName, samplerName, coord2)             SAMPLE_TEXTURE2D_ARRAY(textureName, samplerName, coord2, SLICE_ARRAY_INDEX)
     #define SAMPLE_TEXTURE2D_X_LOD(textureName, samplerName, coord2, lod)    SAMPLE_TEXTURE2D_ARRAY_LOD(textureName, samplerName, coord2, SLICE_ARRAY_INDEX, lod)
     #define GATHER_TEXTURE2D_X(textureName, samplerName, coord2)             GATHER_TEXTURE2D_ARRAY(textureName, samplerName, coord2, SLICE_ARRAY_INDEX)
     #define GATHER_RED_TEXTURE2D_X(textureName, samplerName, coord2)         GATHER_RED_TEXTURE2D(textureName, samplerName, float3(coord2, SLICE_ARRAY_INDEX))
-    #define GATHER_GREEN_TEXTURE2D_X(textureName, samplerName, coord2)       GATHER_GREEN_TEXTURE2D(textureName, samplerName, float3(coord2, SLICE_ARRAY_INDEX))
```
Comment (anisunity) [inline com.unity.render-pipelines.high-definition/Runtime/ShaderLibrary/TextureXR.hlsl]: This is needed, why remove it? @fabien-unity

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipeline.cs (hunk 1 above)
Comment (fabien-unity) [reply com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipeline.cs]: This code will make sure any standard draw call will work with single-pass instancing : it generates 2 mesh instances (one for each eye).
In the ScreenSpaceShadows case, there's a blit that requires it.

File: com.unity.render-pipelines.high-definition/Runtime/ShaderLibrary/TextureXR.hlsl (hunk 2 above)
Comment (fabien-unity) [reply com.unity.render-pipelines.high-definition/Runtime/ShaderLibrary/TextureXR.hlsl]: I've restored the macros and added the missing ones too

**Comment by fabien-unity** [comment]
I've updated this PR with the remaining effects and restarted Katana.

**Comment by anisunity** [comment]
At the moment big regression with this PR, should not be merged!

**Comment by enqui** [comment]
I look forward to testing this.  Please provide a new package build as soon as you got something working properly out of the box for 19.3 Beta... it was very disappointing to see that DXR is not really functional out of the box in the beta version.

**Comment by anisunity** [comment]
Closing the PR FTM

**Pull request closed without merging**